packages/babel-plugin-class-features/src/index.js

```diff
@@ -0,0 +1,189 @@
+import { declare } from "@babel/helper-plugin-utils";
+import nameFunction from "@babel/helper-function-name";
+import { types as t } from "@babel/core";
+import {
+  buildPrivateNamesNodes,
+  buildPrivateNamesMap,
+  transformPrivateNamesUsage,
+  buildFieldsInitNodes,
+} from "./fields";
+import { injectInitialization, extractComputedKeys } from "./misc";
+import {
+  enableFeature,
+  verifyUsedFeatures,
+  FEATURES,
+  setLoose,
+  isLoose,
+} from "./features";
+
+export { enableFeature, FEATURES, setLoose };
+
+const getFeatureOptions = (options, name) => {
+  const value = options[name];
+
+  if (value === undefined || value === false) return { enabled: false };
+  if (value === true) return { enabled: true, loose: false };
+
+  if (typeof value === "object") {
+    if (
+      typeof value.loose !== "undefined" &&
+      typeof value.loose !== "boolean"
+    ) {
+      throw new Error(`.${name}.loose must be a boolean or undefined.`);
+    }
+
+    return { enabled: true, loose: !!value.loose };
+  }
+
+  throw new Error(
+    `.${name} must be a boolean, an object with a 'loose'` +
+      ` property or undefined.`,
+  );
+};
+
+export default declare((api, options) => {
+  api.assertVersion(7);
+
+  const fields = getFeatureOptions(options, "fields");
+  const privateMethods = getFeatureOptions(options, "privateMethods");
+  const decorators = getFeatureOptions(options, "decorators");
+
+  return {
+    manipulateOptions(opts, parserOpts) {
+      if (fields) {
+        parserOpts.plugins.push("classProperties", "classPrivateProperties");
+      }
+    },
+
+    pre() {
```

Comment:
I'm a bit concerned around how some of this will work if there are multiple versions of things. If we eventually add decorators with their own `inherits`, we'll have two separate copies of this plugin running. Would we want to allow both to run? We could potentially have one detect that another instance has already loaded or something, but then we still have the question of _which_ one we end up choosing.

Reply:
For now, only the first plug-in runs (actually both runs, but it is a noop because everything is already transformed).
Probably I should `require("package.json"). version` and only run the most recent version? Or is it overkill?

Reply:
Btw, I will merge the decorators transform before the next release, to check that everything works.

Reply:
The big question for me is if we _want_ plugins like `proposal-class-properties` and `proposal-decorators` at the same time as `babel-plugin-class-features` (which I also notice we may want to be `babel-plugin-proposal-class-features`)?

For instance, we could also deprecate those two plugins, and instead instruct people to use

```
plugins: [
  ["@babel/proposal-class-features", {
    decorators: true,
    publicFields: true,
    privateFields: true,
  }]
]
```
which would more nicely bundle up everything more consistently with our current plugin design.

It's more work for users though, which is why I'm on the fence, but it would mean things are less confusingly side-effectful.

Reply:
My initial idea was to create a single plugin, but different people advocated for keeping the interface as separate plugins.

> The big question for me is if we want plugins like proposal-class-properties and proposal-decorators at the same time as babel-plugin-class-features

Unless people rely on exact versions of the class-features plugin (without `^`) and, at the same time, npm/yarn doesn't dedupe this package, I don't think that using both `class-features` and `class-properties` at the same time would cause confusing effects.
Anyway, we could disallow using them together if you think that it would be better.

> which I also notice we may want to be babel-plugin-proposal-class-features

There will be a time when this plugin will transform standard features and proposals (e.g. when class fields become standard). This was one of the reason that the multi-plugins interface was preferred.

I'm going to split up the actual transform into another PR, so we can keep discussing the interface without blocking https://github.com/babel/babel/issues/8654 and https://github.com/babel/babel/pull/8248

Reply:
> different people advocated for keeping the interface as separate plugins.

Do you have a link to those discussions? I'm concerned that the approach here will make it really hard to maintain and predict how changes might affect users.

> Unless people rely on exact versions of the class-features plugin (without ^) and, at the same time, npm/yarn doesn't dedupe this package, I don't think that using both class-features and class-properties at the same time would cause confusing effects.

Sorry, I should clarify, I don't mean whether they should work together, I mean whether we should maintain both. I'd be tempted to have `babel-plugin-proposal-class-features` be an entire replacement for `proposal-decorators` and `proposal-class-properties`, with those being removed from the monorepo.

> There will be a time when this plugin will transform standard features and proposals (e.g. when class fields become standard). This was one of the reason that the multi-plugins interface was preferred.

There hard part there is that in many cases it's not possible to have one without the other. If we keep things bundled in one plugin, at least it is clear to users that this plugin of class props + decorators + private would likely take recedence, so if say class fields stabilized first, preset-env could default-include

```
['transform-class-features', { publicFields: true }]
```
Which could throw if it encounters classes that have public fields that are decorated, or private fields.

Then if the user adds
```
['proposal-class-features', { decorators: true }]
```
this would essentially override the previous config because it would run first, so it could loudly throw if it encounters public/private fields because they haven't been enabled on that instance of the plugin. Potentially we'd only have to technically error here if a class contained _both_ a decorator and a public field.

Reply:
> Do you have a link to those discussions?

Unfortunately there isn't very much in the meeting notes - the older reference I could find is https://github.com/babel/notes/blob/f58a330d02a2ba7eb29f85dfb88f3c95174d077b/2018/2018-09/sept-10.md#new-class-proposals-decorators-private-fields-etc, which is three months after this PR was opened.

> Sorry, I should clarify, I don't mean whether they should work together, I mean whether we should maintain both.

`class-properties` is only a switch for the `class-features` plugin, it's not something on itself (https://github.com/nicolo-ribaudo/babel/blob/e70d0f03543058ae804bb7814dd7936ae9e0ace7/packages/babel-plugin-proposal-class-properties/src/index.js).

> Then if the user adds
> ```
> ['proposal-class-features', { decorators: true }]
> ```
> this would essentially override the previous config because it would run first, so it could loudly throw if it encounters public/private fields because they haven't been enabled on that instance of the plugin.

I don't think that this is what a user expects: decorators and class fields are different features (even if they need to be implemented together). Why would enabling decorators disable class fields?
Having the separate "switchers" plugin avoids this confusing behavior: if I enable decorators, it doesn't affect class fields being enabled/disabled.

Btw, the fact that you used both `transform` and `proposal` shows that this package's name shouldn't contain either (or should contain both), unless you meant that they should be two separated plugins.

Reply:
If I had to choose, I would disable using the `class-features` plugin directly, since I agree that this configuration would be confusing:

```json
plugins: [
  ["@babel/plugin-class-features", { "classFields": true }],
  ["@babel/plugin-class-features", { "decorators": true }],
]
```

would be surprising if none of them are hidden inside a preset.